flink/v1.17/flink/src/main/java/org/apache/iceberg/flink/source/eventtimeextractor/EventTimeExtractorRecordEmitter.java

```diff
@@ -0,0 +1,76 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.flink.source.eventtimeextractor;
+
+import org.apache.flink.api.common.eventtime.Watermark;
+import org.apache.flink.api.connector.source.SourceOutput;
+import org.apache.iceberg.flink.source.reader.RecordAndPosition;
+import org.apache.iceberg.flink.source.reader.SerializableRecordEmitter;
+import org.apache.iceberg.flink.source.split.IcebergSourceSplit;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Emitter which emits the record with event time and updates the split position.
+ *
+ * <p>The Emitter also emits watermarks at the beginning of every split, and sets the event
+ * timestamp based on the provided {@link IcebergEventTimeExtractor}.
+ */
+public final class EventTimeExtractorRecordEmitter<T> implements SerializableRecordEmitter<T> {
+  private static final Logger LOG = LoggerFactory.getLogger(EventTimeExtractorRecordEmitter.class);
+  private final IcebergEventTimeExtractor timeExtractor;
+  private String lastSplit = null;
+  private long watermark;
+
+  public EventTimeExtractorRecordEmitter(IcebergEventTimeExtractor timeExtractor) {
+    this.timeExtractor = timeExtractor;
+  }
+
+  @Override
+  public void emitRecord(
+      RecordAndPosition<T> element, SourceOutput<T> output, IcebergSourceSplit split) {
+    if (!split.splitId().equals(lastSplit)) {
+      long newWatermark = timeExtractor.extractWatermark(split);
+      if (newWatermark < watermark) {
+        LOG.warn(
+            "Watermark decreased. PreviousWM {}, currentWM {}, previousSplit {}, currentSplit {}.",
+            watermark,
+            newWatermark,
+            lastSplit,
+            split.splitId());
+      }
+      watermark = newWatermark;
+      output.emitWatermark(new Watermark(watermark));
+      lastSplit = split.splitId();
+    }
+
+    long eventTime = timeExtractor.extractEventTime(element.record());
+    if (eventTime <= watermark) {
+      LOG.warn(
+          "Late event arrived. PreviousWM {}, split {}, eventTime {}, record {}.",
+          watermark,
+          split,
+          eventTime,
+          element.record());
+    }
+
+    output.collect(element.record(), eventTime);
```

Comment:
I actually don't think Iceberg source should use the `collect` method, as Iceberg record/row doesn't carry inherent timestamp metadata (like Kafka records may). Iceberg source only needs the split timestamp/watermark extractor.

```
    /**
     * Emit a record with a timestamp.
     *
     * <p>Use this method if the source system has timestamps attached to records. Typical examples
     * would be Logs, PubSubs, or Message Queues, like Kafka or Kinesis, which store a timestamp
     * with each event.
     *
     * <p>The events typically still pass through a {@link TimestampAssigner}, which may decide to
     * either use this source-provided timestamp, or replace it with a timestamp stored within the
     * event (for example if the event was a JSON object one could configure aTimestampAssigner that
     * extracts one of the object's fields and uses that as a timestamp).
     *
     * @param record the record to emit.
     * @param timestamp the timestamp of the record.
     */
    void collect(T record, long timestamp);
```

Reply:
Are you suggesting, that we should use `output.collect(element.record());` instead of `output.collect(element.record(), eventTime);`?

If we use `output.collect(element.record());` then it will fall back to:
```
    @Override
    public final void collect(T record) {
        collect(record, TimestampAssigner.NO_TIMESTAMP);
    }
```

This would mean that any time we try to use the record timestamp for something, like throttling, then we will face the following exception:
```
(f70b076ad52ece48837ec5d43245d956_20ba6b65f97481d5570070de90e4e791_0_0) switched from RUNNING to FAILED with failure cause:
java.lang.RuntimeException: Record has Long.MIN_VALUE timestamp (= no timestamp marker). Is the time characteristic set to 'ProcessingTime', or did you forget to call 'DataStream.assignTimestampsAndWatermarks(...)'?
	at org.apache.flink.streaming.api.windowing.assigners.TumblingEventTimeWindows.assignWindows(TumblingEventTimeWindows.java:83)
	at org.apache.flink.streaming.runtime.operators.windowing.WindowOperator.processElement(WindowOperator.java:284)
	at org.apache.flink.streaming.runtime.io.RecordProcessorUtils.lambda$getRecordProcessor$0(RecordProcessorUtils.java:60)
	at org.apache.flink.streaming.runtime.tasks.OneInputStreamTask$StreamTaskNetworkOutput.emitRecord(OneInputStreamTask.java:237)
	at org.apache.flink.streaming.runtime.io.AbstractStreamTaskNetworkInput.processElement(AbstractStreamTaskNetworkInput.java:146)
	at org.apache.flink.streaming.runtime.io.AbstractStreamTaskNetworkInput.emitNext(AbstractStreamTaskNetworkInput.java:110)
	at org.apache.flink.streaming.runtime.io.StreamOneInputProcessor.processInput(StreamOneInputProcessor.java:65)
	at org.apache.flink.streaming.runtime.tasks.StreamTask.processInput(StreamTask.java:550)
	at org.apache.flink.streaming.runtime.tasks.mailbox.MailboxProcessor.runMailboxLoop(MailboxProcessor.java:231)
	at org.apache.flink.streaming.runtime.tasks.StreamTask.runMailboxLoop(StreamTask.java:839)
	at org.apache.flink.streaming.runtime.tasks.StreamTask.invoke(StreamTask.java:788)
	at org.apache.flink.runtime.taskmanager.Task.runWithSystemExitMonitoring(Task.java:952)
	at org.apache.flink.runtime.taskmanager.Task.restoreAndInvoke(Task.java:931)
	at org.apache.flink.runtime.taskmanager.Task.doRun(Task.java:745)
	at org.apache.flink.runtime.taskmanager.Task.run(Task.java:562)
	at java.base/java.lang.Thread.run(Thread.java:829)
```

So we need a way to extract the timestamp from the record to effectively use watermarks. This is one of the reasons why we used a single interface for watermark and timestamp extraction. There is not much use of one without the other.

Reply:
> Are you suggesting, that we should use output.collect(element.record()); instead of output.collect(element.record(), eventTime);?

Yes, that is what I am suggesting. Looking at the Javadoc, `collect(T record, long timestamp)` is mainly for ` Kafka or Kinesis, which store a timestamp with each event.`

In Kafka case, the timestamp is extracted from Kafka `ConsumerRecord` metadata field, which can be set by producer or broker. But we don't have such equivalent per-record timestamp metadata in Iceberg. 
https://github.com/stevenzwu/flink-connector-kafka/blob/main/flink-connector-kafka/src/main/java/org/apache/flink/connector/kafka/source/reader/KafkaRecordEmitter.java#L52

Regarding the exception, this is sth I don't fully understand. The emitter already emits watermark, that is not enough for source watermark alignment?

The exception was throw from tumbling window operator, which requires the event timestamp per record. watermark strategy  can be [set and overridden after the source](https://nightlies.apache.org/flink/flink-docs-release-1.17/docs/dev/datastream/event-time/generating_watermarks/#using-watermark-strategies) or can be [passed into the source directly](https://nightlies.apache.org/flink/flink-docs-release-1.17/docs/dev/datastream/event-time/generating_watermarks/#watermark-strategies-and-the-kafka-connector).

Reply:
Thanks for the info!
Finally I was able to make it work without our own timestamp extractor interaface!!!

Reply:
> it is unclear to me if the watermark strategy is passed into the source operator (like Iceberg), how does it interfere/live with the watermark emitted by the emitter at split level.

@pvary this is the part I don't know. the source watermark alignment works with emitter emitting split level watermark values that should advance watermark. but then if the watermark strategy was passed into the source operator directly. how are the watermark actually working especially related to the watermark alignment in the source reader.

Reply:
I did not see anything in the code to handle the situation specifically.
I expect that adding a watermark strategy will cause emitting watermarks from there as well, which would result in a mess with regards to the watermarks. I would suggest to use the `watermarkExtractor` with `noWatermarks` every time.